src/devices/bus/spectrum/zxbus.h

```diff
@@ -81,24 +81,37 @@ DECLARE_DEVICE_TYPE(ZXBUS_SLOT, zxbus_slot_device)
 
 
 class zxbus_device : public device_t
+                   , public device_memory_interface
 {
 public:
 	zxbus_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
 
 	template <typename T> void set_iospace(T &&tag, int spacenum) { m_iospace.set_tag(std::forward<T>(tag), spacenum); }
-	template<typename T> void install_device(offs_t addrstart, offs_t addrend, T &device, void (T::*map)(class address_map &map), uint64_t unitmask = ~u64(0))
+
+	template<typename T> void install_device(offs_t addrstart, offs_t addrend, T &device, void (T::*map)(class address_map &map), u64 unitmask = ~u64(0))
 	{
 		m_iospace->install_device(addrstart, addrend, device, map, unitmask);
 	}
 
+	u8 io_r(offs_t offset);
+	void io_w(offs_t offset, u8 data);
+
 	void add_slot(zxbus_slot_device &slot);
+	void set_custom_spaces() { m_allocspaces = true; }
+
+	space_config_vector memory_space_config() const override;
+	const address_space_config m_io_config;
 
 protected:
 	zxbus_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock);
 
-	virtual void device_start() override;
+	void device_config_complete() override;
+	void device_start() override;
```

Comment:
Please include the `virtual` qualifier on overriden virtual member functions.  The vast majority of MAME includes it.